rs/sns/root/src/install_index_canister_if_missing.rs

```diff
@@ -0,0 +1,138 @@
+use crate::{logs::ERROR, pb::v1::SnsRootCanister};
+
+use candid::Encode;
+use ic_base_types::{CanisterId, PrincipalId};
+use ic_canister_log::log;
+use ic_icrc1_index_ng::{IndexArg, InitArg};
+use ic_management_canister_types_private::CanisterInstallMode;
+use ic_nervous_system_clients::canister_id_record::CanisterIdRecord;
+use ic_nervous_system_root::change_canister::ChangeCanisterRequest;
+use ic_nervous_system_runtime::{CdkRuntime, Runtime};
+use ic_nns_constants::SNS_WASM_CANISTER_ID;
+use ic_sns_governance_api::pb::v1::{GetRunningSnsVersionRequest, GetRunningSnsVersionResponse};
+use std::cell::RefCell;
+use std::thread::LocalKey;
+
+#[derive(candid::CandidType, candid::Deserialize)]
+struct GetWasmRequest {
+    hash: Vec<u8>,
+}
+
+#[derive(candid::CandidType, candid::Deserialize)]
+struct GetWasmResponse {
+    wasm: Option<SnsWasm>,
+}
+
+#[derive(candid::CandidType, candid::Deserialize)]
+struct SnsWasm {
+    wasm: Vec<u8>,
+    canister_type: i32,
+    proposal_id: Option<u64>,
+}
+
+impl SnsRootCanister {
+    /// Reinstalls the index canister if it's needed (i.e., if it has no wasm module).
+    pub async fn install_index_canister_if_missing(self_ref: &'static LocalKey<RefCell<Self>>) {
+        let (governance_canister_id, ledger_canister_id, index_canister_id) = self_ref.with(|r| {
+            let r = r.borrow();
+            (
+                r.governance_canister_id,
+                r.ledger_canister_id,
+                r.index_canister_id,
+            )
+        });
+
+        if !should_install_index_canister(index_canister_id).await {
+            return;
+        }
+
+        let result = try_install_index_canister(
+            governance_canister_id,
+            ledger_canister_id,
+            index_canister_id,
+        )
+        .await;
+
+        if let Err(e) = result {
+            log!(ERROR, "Error installing index canister: {}", e);
+        }
+    }
+}
+
+fn get_canister_id(principal_id: Option<PrincipalId>, label: &str) -> Result<CanisterId, String> {
+    let principal_id = principal_id.ok_or(format!("No canister id for {label} provided"))?;
+    CanisterId::try_from_principal_id(principal_id)
+        .map_err(|e| format!("Error getting canister id for {label}: {e}"))
+}
+
+async fn should_install_index_canister(index_canister_id: Option<PrincipalId>) -> bool {
+    let Ok(index_canister_id) = get_canister_id(index_canister_id, "Index") else {
```

Comment:
I think we want to really restrict this to the one known messed up canister. To do that, this would be replaced with something like

```
if index_canister_id != Some(ICVC_INDEX_CANISTER_ID) {
```